hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecord.java

```diff
@@ -40,11 +62,192 @@ public HoodieRecord<T> newInstance() {
     return new HoodieAvroRecord<>(this);
   }
 
+  @Override
+  public HoodieRecord<T> newInstance(HoodieKey key, HoodieOperation op) {
+    return new HoodieAvroRecord<>(key, data, op);
+  }
+
+  @Override
+  public HoodieRecord<T> newInstance(HoodieKey key) {
+    return new HoodieAvroRecord<>(key, data);
+  }
+
   @Override
   public T getData() {
     if (data == null) {
       throw new IllegalStateException("Payload already deflated for record.");
     }
     return data;
   }
+
+  @Override
+  public String getRecordKey(Option<BaseKeyGenerator> keyGeneratorOpt) {
+    return getRecordKey();
+  }
+
+  @Override
+  public Comparable<?> getOrderingValue() {
+    return data.getOrderingValue();
+  }
+
+  @Override
+  public Option<IndexedRecord> toIndexedRecord(Schema schema, Properties prop) throws IOException {
+    return getData().getInsertValue(schema, prop);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  //
+  // NOTE: This method duplicates those ones of the HoodieRecordPayload and are placed here
+  //       for the duration of RFC-46 implementation, until migration off `HoodieRecordPayload`
+  //       is complete
+  //
+  // TODO cleanup
+
+  // NOTE: This method is assuming semantic that `preCombine` operation is bound to pick one or the other
+  //       object, and may not create a new one
+  @Override
+  public HoodieRecord<T> preCombine(HoodieRecord<T> previousRecord) {
+    T picked = unsafeCast(getData().preCombine(previousRecord.getData()));
+    if (picked instanceof HoodieMetadataPayload) {
```

Comment:
such an easter egg. cc @codope @yihua to file a JIRA to remove this bandaid if inherited.

Reply:
https://issues.apache.org/jira/browse/HUDI-4222